src/compiler/checker.ts

```diff
@@ -19860,7 +19860,7 @@ namespace ts {
                             return varianceResult;
                         }
                     }
-                    else if (isReadonlyArrayType(target) ? isArrayOrTupleType(source) : isArrayType(target) && isTupleType(source) && !source.target.readonly) {
+                    else if (isReadonlyArrayType(target) ? everyType(source, isArrayOrTupleType) : isArrayType(target) && isTupleType(source) && !source.target.readonly) {

```

Comment:
There is a good chance that this should do the same thing in the `whenFalse` branch of this ternary expression:
```suggestion
                    else if (isReadonlyArrayType(target) ? everyType(source, isArrayOrTupleType) : isArrayType(target) && everyType(source, isTupleType) && !source.target.readonly) {
```

but to make such a change I will have to figure out a test case validating it. I don't think it's going to be super hard but my available time is limited rn.

Reply:
The false branch is just for read-only types, so yeah, it should *probably* be the same, otherwise it'll work for non-readonly tuples, but not readonly ones.

Reply:
I've pushed out a fix and tests for this.